scs2-definition/src/main/java/us/ihmc/scs2/definition/robot/urdf/URDFTools.java

```diff
@@ -209,66 +228,261 @@ public static URDFModel loadURDFModel(InputStream inputStream, Collection<String
+      {
+         return model1;
+      }
+
+      // Get all names of links in model1
+      List<String> model1LinkNames = new ArrayList<>();
+      List<String> model2LinkNames = new ArrayList<>();
+
+      if (model1.getLinks() != null)
+      {
+         for (URDFLink link : model1.getLinks())
          {
-            InputSource is = new InputSource(inputStream);
-            SAXParserFactory sax = SAXParserFactory.newInstance();
-            sax.setNamespaceAware(false);
-            XMLReader reader;
+            model1LinkNames.add(link.getName());
+         }
+      }
 
-            try
+      if (model2.getLinks() != null)
+      {
+         for (URDFLink link : model2.getLinks())
+         {
+            model2LinkNames.add(link.getName());
+         }
+      }
+
+      if (model2.getJoints() != null)
+      {
+         for (URDFJoint joint : model2.getJoints())
+         {
+            // Check if the model1 has any links that are referenced in model2
+            if (model1LinkNames.contains(joint.getParent().getLink()))
             {
-               reader = sax.newSAXParser().getXMLReader();
+               return mergeChildIntoParentModel(model1, model2);
             }
-            catch (SAXException | ParserConfigurationException e)
+            if (model1LinkNames.contains(joint.getChild().getLink()))
             {
-               throw new JAXBException(e);
+               return mergeChildIntoParentModel(model2, model1);
             }
+         }
+      }
 
-            SAXSource source = new SAXSource(reader, is);
-            urdfModel = (URDFModel) um.unmarshal(source);
+      if (model1.getJoints() != null)
+      {
+         for (URDFJoint joint : model1.getJoints())
+         {
+            // Check if the model2 has any links that are referenced in model1
+            if (model2LinkNames.contains(joint.getParent().getLink()))
+            {
+               return mergeChildIntoParentModel(model2, model1);
+            }
+            if (model2LinkNames.contains(joint.getChild().getLink()))
+            {
+               return mergeChildIntoParentModel(model1, model2);
+            }
          }
-         resolvePaths(urdfModel, allResourceDirectories, resourceClassLoader);
+      }
 
-         if (!parserProperties.linksToIgnore.isEmpty() && urdfModel.getLinks() != null)
-            urdfModel.getLinks().removeIf(urdfLink -> parserProperties.linksToIgnore.contains(urdfLink.getName()));
-         if (!parserProperties.jointsToIgnore.isEmpty() && urdfModel.getJoints() != null)
-            urdfModel.getJoints().removeIf(urdfJoint -> parserProperties.jointsToIgnore.contains(urdfJoint.getName()));
-         if (!parserProperties.parseSensors)
-            urdfModel.getGazebos().removeIf(gazebo -> gazebo.getSensor() != null);
+      // If the model1 doesn't have any links that are in the model2, we define the model2 to be the parent
+      return mergeChildIntoParentModel(model2, model1);
```

Comment:
I'd suggest inlining the `mergeChildIntoParentModel` method in here, as it now relies solving the parent-child relationship outside making it vulnerable to idiot bugs 😅
Above, I'd make 2 new local variables `URDFModel parentModel` and `URDFModel childModel` and resolve them, then after you should be able to inline the `mergeChildIntoParentModel` only once at the end of this method.

Reply:
True, that was able to happen, just had to set the parent or child model to the correct model and after that has been done, at the end collect all the joints from the parent and child.